sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -121,6 +132,19 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
         if canImplicitlyCast(fromExp, toType, literalType) =>
       simplifyNumericComparison(be, fromExp, toType, value)
 
+    case in @ In(Cast(fromExp, toType: NumericType, _), list)
+      if list.forall(v =>
+        v.isInstanceOf[Literal] && canImplicitlyCast(fromExp, toType, v.dataType)
+      ) =>
+      val (newValueList, exp) =
+        list.map(lit => unwrapCast(EqualTo(in.value, lit)))
+          .partition {
+            case EqualTo(_, _: Literal) => true
+            case And(IsNull(_), Literal(null, BooleanType)) => false
+          }
+      val unwrapIn = In(fromExp, newValueList.map {case EqualTo(_, lit) => lit})
+      (exp :+ unwrapIn).reduce(Or)
```

Comment:
nit: for the `exp` part, maybe we could convert the list into a single value `And(IsNull(_), Literal(null, BooleanType))`? since they are all the same.